source/agora/consensus/Quorum.d

```diff
@@ -0,0 +1,1229 @@
+        number of quorum configuration permutations.
+
+    - countNodeInclusions(): it counts the number of times a node
+        has been included in other nodes' quorum sets. Using this
+        makes it easy to determine if the algorithm behaves correctly
+        when the staken amounts are differen between nodes.
+
+    Note: If you use AscendingQuadratic for too many nodes, it will
+    assert as you will quickly run out of BOA.
+
+    Note: Intersection checks take roughly ~2 minutes for a single configuration
+    of 16 node quorums, by default we disable these checks for many nodes
+    unless overriden with -version=EnableAllIntersectionChecks.
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.consensus.Quorum;
+
+/// Force intersection checks on all unittests
+//version = EnableAllIntersectionChecks;
+
+import agora.common.Amount;
+import agora.common.BitField;
+import agora.common.Config;
+import agora.common.crypto.Key;
+import agora.common.Hash;
+import agora.common.Set;
+import agora.consensus.data.Enrollment;
+import agora.consensus.data.UTXOSet;
+import agora.consensus.EnrollmentManager;
+import agora.utils.PrettyPrinter;
+
+import scpd.Cpp;
+import scpd.types.Stellar_SCP;
+import scpd.types.Stellar_types : StellarHash = Hash, NodeID;
+import scpd.types.Utils;
+import scpd.types.XDRBase;
+
+import scpd.quorum.QuorumIntersectionChecker;
+import scpd.quorum.QuorumTracker;
+
+import std.algorithm;
+import std.array;
+import std.conv;
+import std.format;
+import std.math;
+import std.random;
+import std.range;
+import std.string;
+import std.typecons;
+
+import std.stdio;
+
+version (unittest)
+{
+    import std.conv;
+}
+
+/// limits the number of nodes in a quorum set
+private enum MAX_NODES_IN_QUORUM = 7;
+
+/*******************************************************************************
+
+    Build the quorum configuration for the given node key and the registered
+    enrollments. The random seed is used to shuffle the quorum config.
+
+    Params:
+        node_key = the key of the node
+        enrolls = the array of registered enrollments
+        finder = the delegate to find UTXOs with
+        rand_seed = the source of randomness
+
+    Returns:
+        the map of all quorum configurations
+
+*******************************************************************************/
+
+public QuorumConfig buildQuorumConfig ( PublicKey node_key,
+    Enrollment[] enrolls, UTXOFinder finder, Hash rand_seed,
+    size_t line = __LINE__ )
+{
+    Amount[PublicKey] all_stakes = buildStakes(enrolls, finder);
+    NodeStake[] stakes_by_price = orderStakesDescending(all_stakes);
+
+    const Amount min_quorum_amount = Amount(
+        cast(ulong)(10_000_000 * (0.67 *  // todo: add multiply() support
```

Comment:
I'm going to have to remove this piece of logic. There's two reasons:

- Floating point math could lead to inconsistent results across platforms. We need consistency here. Perhaps we could use fixed-point math, but it would require looking for a good library implementation.
- It doesn't entirely make sense that we're adding nodes into a quorum to make that quorum's monetary value 67% of the network, while at the same time having a 67% threshold value for those included nodes. That would actually turn into 44% with the threshold taken into account. Also, within that quorum set SCP treats every node as equal. This means nodes with 1/10th of the quorum stake and those with 9/10th of the value are treated the same. So monetary value != political (voting) power, as it should be. Finally, the node selection logic already makes sure that nodes with a higher stake are more frequently included in a quorum, so the `min_quorum_amount` is unnecessary.

Reply:
There is a standard for double-precision that is IEEE 754. Modern 32-bit computers implements double-precision with two 32-bit memory words and a 64-bit computer can do it with its 64-bit memory word. So We don't need to worry about the output of the floating-point operation.